### PR TITLE
Fixes #4515 - Add support for dynamic bindings

### PR DESCRIPTION
**Pull request by mbacovsky**
I reworked PR https://github.com/theforeman/foreman/pull/1157
Most of the functionality was moved to Apipie (https://github.com/Apipie/apipie-rails/pull/215)
In Foreman we just use the middleware. The headers are now called `Apipie-Checksum` and are added just to calls to `/api` and `/apidoc`.

I fixed controller matcher setup so that it can be extended by plugins.

### CODE REVIEW

**Comment by domcleal** [comment]
@lzap was concerned about the speed of this before, do you have any further thoughts?  Perhaps we can use the Rails.cache in places (especially in production)?

There are test failures on Foreman as it's adding a route but no permission is associated with it (I guess it should be in the default "anonymous" permissions).  Some other errors on Katello (@mccun934).

File: Gemfile [hunk 1]
```diff
@@ -14,7 +14,7 @@ gem "ancestry", "~> 2.0"
 gem 'scoped_search', '>= 2.6.2'
 gem 'net-ldap'
 gem 'uuidtools'
-gem "apipie-rails", "~> 0.0.23"
+gem "apipie-rails", "~> 0.1.0"
```
Comment (domcleal) [inline Gemfile]: Could you also update the spec file please?  (Probably needs both a >= and < requires.)

**Comment by lzap** [comment]
I don't think this should be named "Checksum", more than that I think simple timestamp is enough. But this works for me as long as we do not slow down _development_ startup.

**Comment by domcleal** [comment]
@lzap we should be careful not to slow production startup down too much either.  Users may notice, but also if Apache or Passenger are spawning new workers in response to requests, it can still slow down responses or at least add server load.

**Comment by dLobatog** [comment]
Curious here, why would this slow down startup time so much?

**Comment by domcleal** [comment]
@eLobato I've not re-tested it, but it was checksumming a lot of API documentation during an initializer, looks like that might still be the case.

@mbacovsky, @iNecas looks like apipie's routes are specified with a 1.9 hash syntax, so Foreman tests are partially failing.

**Comment by iNecas** [comment]
How long is foreman going to support 1.8? Apipie doesn't officially support 1.8 anymore (known as not running tests against it). Should be not hard to add and I will be happy to merge the PR fixing the issues (should not really be much). On the other hand, maybe its time for Foreman to move on and say goodbye to 1.8.7.

**Comment by domcleal** [comment]
We said we'd continue to support it until we move to Rails 4, assuming the effort is reasonable.  I suspect we'll drop it this year.

**Comment by mbacovsky** [comment]
I'll look the ruby 1.8 issues in apipie.  During the startup most of the extra time is spent in loading controllers and parsing doc data. I can do that lazy (on first request) which could releave some load. I'll also check if it is possible to count the checksum form the cache when run with use_cache option.

**Comment by mbacovsky** [comment]
Apipie was updated to support ruby 1.8. The server startup  time is not affected when Apipie use_cache is true, which we use in production by default. 
Main portion of the delay was caused by parsing of the controllers, minor part was the checksum computation. It is necessary to call foreman-rake apipie:cache to generate the cache and do all the parsing every time the documentation changed or plugin extending api was added. The checksum computation is now lazy and is done with first /api request.

**Comment by domcleal** [comment]
@mbacovsky so we need to ensure the rake task is called after installation/upgrade, and in plugin packages which add APIs?

Do the rake task and lazy computation work together, so without the rake task having been run it lazily computes instead?

File: Gemfile [hunk 2]
```diff
@@ -14,7 +14,7 @@ gem "ancestry", "~> 2.0"
 gem 'scoped_search', '>= 2.6.2'
 gem 'net-ldap'
 gem 'uuidtools'
-gem "apipie-rails", "~> 0.0.23"
+gem "apipie-rails", "~> 0.1.1"
```
Comment (domcleal) [inline Gemfile:17]: [fixed] Could you update foreman.spec to match please?

**Comment by mbacovsky** [comment]
The workflow should be: Install/configure/add_plugin(change API) -> foreman_rake apipie:cache -> httpd restart.
If you have use_cache = true and didn't run the rake task, then it makes no sense to compute checksums, because you have no way to serve the apipie json. If the cache is turned off it works as expected. Not sure if this should be suggested just in docs or if we have some place to call the rake automatically. Maybe some check if the cache is generated?
Deps on apipie in Spec were updated.

**Comment by lzap** [comment]
What happens if use_cache is true but I will not run the rake task?

Does it fail nicely?

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

**Comment by domcleal** [comment]
Right, I think I see - so running apipie:cache is mandatory anyway for the CLI to function, as it'll require the apipie JSON.  So in that case, I think we'd need to add this job to the %post in foreman.spec (along with the Debian package's postinstall), to the manual and to plugins etc.

**Comment by mbacovsky** [comment]
@domcleal that makes sense. I'll update the spec here. The debs needs to be fixed in the foreman-packaging repo?

**Comment by domcleal** [comment]
@mbacovsky that's right, foreman-packaging repo against the deb/develop branch.

And regarding @lzap's question, would Hammer get a 404 or similar for the JSON if the cache didn't exist?

**Comment by mbacovsky** [comment]
@domcleal, @lzap if the cache is not generated foreman returns `404` and logs `API doc cache not found for '/apidoc/v2.json'. Perhaps you have forgot to run `rake apipie:cache`

spec was updated, debs are on the way

**Comment by mbacovsky** [comment]
I found out development mode is still broken when we keep in application.rb

``` ruby
    if Rails.env.development?
      config.autoload_once_paths += Dir["#{config.root}/app/services/foreman"]
      config.autoload_once_paths += Dir["#{config.root}/app/services/menu"]
    end
```

Investigating...

**Comment by domcleal** [comment]
Yeah, that's http://projects.theforeman.org/issues/3988 and is getting really irritating.  Unfortunately it was added to stop the menu reloading in development mode and losing the registered plugins, so we need to be careful when fixing it that we don't revert that.

**Comment by mbacovsky** [comment]
@iNecas suggested solution that I included in this PR. Seems that it works, but I didn't test the loosing of pliugins yet.

**Comment by mbacovsky** [comment]
Here is the related debian patch provided by @tstrachota   https://github.com/theforeman/foreman-packaging/pull/197

**Comment by iNecas** [comment]
The proper fix for the reloading issue should be here: https://github.com/mbacovsky/foreman/pull/1

Tested with katello as a plugin

**Comment by domcleal** [comment]
Tests fine, thanks.  Just waiting on inclusion of @iNecas' fix, though I added a comment on that sub-PR.

**Comment by mbacovsky** [comment]
@iNecas ' fix is merged in, thanks @iNecas. tests are green

**Comment by domcleal** [comment]
Merged as 70ce8fe and 0527e75 for Foreman 1.5.0.  Thanks @mbacovsky and @iNecas!

**Pull request closed without merging**